ext-apiserver/pkg/controller/cluster/controller_test.go

```diff
@@ -15,77 +14,78 @@ See the License for the specific language governing permissions and
 
-	BeforeEach(func() {
-		instance = Cluster{}
-		instance.Name = "instance-1"
-		expectedKey = "cluster-controller-test-handler/instance-1"
-	})
-
-	AfterEach(func() {
-		client.Delete(instance.Name, &metav1.DeleteOptions{})
-	})
-
-	Describe("when creating a new object", func() {
-		It("invoke the reconcile method", func() {
-			client = cs.ClusterV1alpha1().Clusters("cluster-controller-test-handler")
-			before = make(chan struct{})
-			after = make(chan struct{})
-
-			actualKey := ""
-			var actualErr error = nil
-
-			// Setup test callbacks to be called when the message is reconciled
-			controller.BeforeReconcile = func(key string) {
-				defer close(before)
-				actualKey = key
-			}
-			controller.AfterReconcile = func(key string, err error) {
-				defer close(after)
-				actualKey = key
-				actualErr = err
-			}
-
-			// Create an instance
-			_, err := client.Create(&instance)
-			Expect(err).ShouldNot(HaveOccurred())
-
-			// Verify reconcile function is called against the correct key
-			select {
-			case <-before:
-				Expect(actualKey).To(Equal(expectedKey))
-				Expect(actualErr).ShouldNot(HaveOccurred())
-			case <-time.After(time.Second * 2):
-				Fail("reconcile never called")
-			}
-
-			select {
-			case <-after:
-				Expect(actualKey).To(Equal(expectedKey))
-				Expect(actualErr).ShouldNot(HaveOccurred())
-			case <-time.After(time.Second * 2):
-				Fail("reconcile never finished")
-			}
-		})
-	})
-})
+	instance = Cluster{}
+	instance.Name = "instance-1"
+	expectedKey = "cluster-controller-test-handler/instance-1"
+
+	// When creating a new object, it should invoke the reconcile method.
+	client = cs.ClusterV1alpha1().Clusters("cluster-controller-test-handler")
+	before = make(chan struct{})
+	after = make(chan struct{})
+
+	actualKey := ""
+	var actualErr error = nil
+
+	// Setup test callbacks to be called when the message is reconciled
+	controller.BeforeReconcile = func(key string) {
+		defer close(before)
+		actualKey = key
+	}
+	controller.AfterReconcile = func(key string, err error) {
+		defer close(after)
+		actualKey = key
+		actualErr = err
+	}
+
+	// Create an instance
+	_, err := client.Create(&instance)
+	assertNoError(t, err)
+	// Verify reconcile function is called against the correct key
+	select {
+	case <-before:
+		if !reflect.DeepEqual(actualKey, expectedKey) {
+			t.Fatalf(
+				"Reconcile function was not called with the correct key.\nActual:\t%+v\nExpected:\t%+v",
+				actualKey, expectedKey)
+		}
+		assertNoError(t, actualErr)
+	case <-time.After(time.Second * 2):
+		t.Fatalf("reconcile never called")
+	}
+
+	select {
+	case <-after:
+		if !reflect.DeepEqual(actualKey, expectedKey) {
+			t.Fatalf(
+				"Reconcile function was not called with the correct key.\nActual:\t%+v\nExpected:\t%+v",
+				actualKey, expectedKey)
+		}
+		assertNoError(t, actualErr)
+	case <-time.After(time.Second * 2):
+		t.Fatalf("reconcile never finished")
+	}
+
+	client.Delete(instance.Name, &metav1.DeleteOptions{})
```

Comment:
This should be defered just after the error check on create to make sure executed even when the test hits a fatal error.